File: ellekit/Exception.swift
//
//  Exception.swift
//  ellekit
//
//  Created by charlotte on 2022-11-03.
//

import Foundation
import Darwin

let ARM_THREAD_STATE64_COUNT = MemoryLayout<arm_thread_state64_t>.size/MemoryLayout<UInt32>.size

public func registerEXCPort() {
    
    var exceptionPort = mach_port_t()
        
    mach_port_allocate(mach_task_self_, MACH_PORT_RIGHT_RECEIVE, &exceptionPort)
    mach_port_insert_right(mach_task_self_, exceptionPort, exceptionPort, mach_msg_type_name_t(MACH_MSG_TYPE_MAKE_SEND))
    
    task_set_exception_ports(mach_task_self_, exception_mask_t(EXC_MASK_BREAKPOINT), exceptionPort, EXCEPTION_DEFAULT, ARM_THREAD_STATE64)
                
    DispatchQueue(label: "exceptionHandler", qos: .userInteractive).async {
        while true {
            let head = UnsafeMutablePointer<mach_msg_header_t>.allocate(capacity: 0x4000)
            
            defer { head.deallocate() }
            
            var ret = mach_msg(head,
                               MACH_RCV_MSG | MACH_RCV_LARGE | Int32(MACH_MSG_TIMEOUT_NONE),
                               0,
                               0x4000,
                               exceptionPort,
                               0, 0)
            guard ret == KERN_SUCCESS else {
                print("[-] error receiving from port:", mach_error_string(ret) ?? "")
                continue
            }
            
            let req = head.withMemoryRebound(to: exception_raise_request.self,
                                             capacity: 0x4000) { $0.pointee }
            
            let thread_port = req.thread.name
            let task_port = req.task.name
                        
            defer {
                var reply = exception_raise_reply()
                reply.Head.msgh_bits = req.Head.msgh_bits & UInt32(MACH_MSGH_BITS_REMOTE_MASK)
                reply.Head.msgh_size = mach_msg_size_t(MemoryLayout.size(ofValue: reply))
                reply.Head.msgh_remote_port = req.Head.msgh_remote_port
                reply.Head.msgh_local_port = mach_port_t(MACH_PORT_NULL)
                reply.Head.msgh_id = req.Head.msgh_id + 0x64
                
                reply.NDR = req.NDR
                reply.RetCode = KERN_SUCCESS
                
                ret = mach_msg(&reply.Head,
                               1,
                               reply.Head.msgh_size,
                               0,
                               mach_port_name_t(MACH_PORT_NULL),
                               MACH_MSG_TIMEOUT_NONE,
                               mach_port_name_t(MACH_PORT_NULL))
                mach_port_deallocate(mach_task_self_, thread_port)
                mach_port_deallocate(mach_task_self_, task_port)
                
                if ret != KERN_SUCCESS {
                    print("[-] error sending reply to exception: ", mach_error_string(ret) ?? "")
                }
            }
            
            var state = arm_thread_state64()
            var stateCnt = mach_msg_type_number_t(ARM_THREAD_STATE64_COUNT)
            
            ret = withUnsafeMutablePointer(to: &state) {
                $0.withMemoryRebound(to: UInt32.self, capacity: MemoryLayout<arm_thread_state64>.size) {
                    thread_get_state(thread_port, ARM_THREAD_STATE64, $0, &stateCnt)
                }
            }
            
            let formerPtr = UnsafeMutableRawPointer(bitPattern: UInt(state.__pc))!
                        
            print(state)
            
            if let newPtr = hooks[formerPtr] {
                print("[+] changed pc to", newPtr)
                state.__pc = UInt64(UInt(bitPattern: newPtr))
            }
                                
            ret = withUnsafeMutablePointer(to: &state, {
                $0.withMemoryRebound(to: UInt32.self, capacity: MemoryLayout<arm_thread_state64>.size, {
                    thread_set_state(thread_port, ARM_THREAD_STATE64, $0, mach_msg_type_number_t(ARM_THREAD_STATE64_COUNT))
                })
            })
            
            thread_resume(thread_port)
                        
            guard ret == KERN_SUCCESS else {
                print("[-] error getting thread state:", mach_error_string(ret) ?? "")
                continue
            }
        }
    }
}

File: ellekit/Exception.c
//
//  Exception.c
//  ellekit
//
//  Created by charlotte on 2022-11-03.
//

#include <mach/mach.h>
#include <pthread/pthread.h>
#include <stdio.h>

static mach_port_name_t myExceptionPort;

void *exc_handler(void *ignored)
{
    // Exception handler – runs a message loop. Refactored into a standalone function
    // so as to allow easy insertion into a thread (can be in same program or different)
    mach_msg_return_t rc;
    fprintf(stderr, "Exc handler listening\n");
    // The exception message, straight from mach/exc.defs (following MIG processing) // copied here for ease of reference.
    typedef struct {
        mach_msg_header_t Head;
        /* start of the kernel processed data */
        mach_msg_body_t msgh_body;
        mach_msg_port_descriptor_t thread;
        mach_msg_port_descriptor_t task;
        /* end of the kernel processed data */
        NDR_record_t NDR;
        exception_type_t exception;
        mach_msg_type_number_t codeCnt;
        integer_t code[2];
        int flavor;
        mach_msg_type_number_t old_stateCnt;
        natural_t old_state[144];
    } Request;

    Request exc;

    struct rep_msg {
        mach_msg_header_t Head;
        NDR_record_t NDR;
        kern_return_t RetCode;
    } rep_msg;


    for(;;) {
        rc = mach_msg(&exc.Head,
                      MACH_RCV_MSG|MACH_RCV_LARGE,
                      0,
                      sizeof(Request),
                      myExceptionPort, // Remember this was global – that's why.
                      MACH_MSG_TIMEOUT_NONE,
                      MACH_PORT_NULL);

        if(rc != MACH_MSG_SUCCESS) {
            /*... */
            break ;
        };


        // Normally we would call exc_server or other. In this example, however, we wish
        // to demonstrate the message contents:

        printf("Got message %u. Exception : %u Flavor: %u. Code %u/%u. State count is %u\n" ,
               exc.Head.msgh_id,
               exc.exception,
               exc.flavor,
               exc.code[0],
               exc.code[1],
               exc.old_stateCnt);
        
        int i;

        for (i = 0; i < exc.old_stateCnt; ++i)
        {
          printf("%d ", exc.old_state[i]);
        }

        rep_msg.Head = exc.Head;
        rep_msg.NDR = exc.NDR;
        rep_msg.RetCode = KERN_FAILURE;

        printf("### handled mach exception");

        kern_return_t result;
        if (rc == MACH_MSG_SUCCESS) {
            result = mach_msg(&rep_msg.Head,
                              MACH_SEND_MSG,
                              sizeof (rep_msg),
                              0,
                              MACH_PORT_NULL,
                              MACH_MSG_TIMEOUT_NONE,
                              MACH_PORT_NULL);
        }
        
        __asm__("mov x0, #12");
        __asm__("mov x16, #1");
        __asm__("svc #0x80");
    }

    return NULL;

} // end exc_handler

void catchMachExceptions(void) {

    kern_return_t rc = 0;
    exception_mask_t excMask = EXC_MASK_BAD_ACCESS | EXC_MASK_BAD_INSTRUCTION | EXC_MASK_ARITHMETIC | EXC_MASK_SOFTWARE | EXC_MASK_BREAKPOINT;

    rc = mach_port_allocate(mach_task_self(), MACH_PORT_RIGHT_RECEIVE, &myExceptionPort);
    if (rc != KERN_SUCCESS) {
        fprintf(stderr, "[-] Fail to allocate exception port\n");
        return;
    }

    rc = mach_port_insert_right(mach_task_self(), myExceptionPort, myExceptionPort, MACH_MSG_TYPE_MAKE_SEND);
    if (rc != KERN_SUCCESS) {
        fprintf(stderr, "[-] Fail to insert right\n");
        return;
    }

    rc = thread_set_exception_ports(mach_thread_self(), excMask, myExceptionPort, EXCEPTION_DEFAULT, MACHINE_THREAD_STATE);
    if (rc != KERN_SUCCESS) {
        fprintf(stderr, "[-] Fail to set exception\n");
        return;
    }

    // at the end of catchMachExceptions, spawn the exception handling thread
    pthread_t thread;
    pthread_create(&thread, NULL, exc_handler, NULL);
}

int testExceptionPort(void) {
    __asm__("brk #125");
    return 0;
}
